tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/clientserver/Checker.java

```diff
@@ -82,41 +85,38 @@ public Checker(int sleepMillis, TestData td, int fails) {
             while ((now > max) && !td.getMaxSimultaneousWorkers().compareAndSet(max, now)) {
                 max = td.getMaxSimultaneousWorkers().get();
             }
+
             if (fails > 0) {
-                Thread.sleep(millis / 2);
                 fails--;
-                RuntimeErrorException e = new RuntimeErrorException(new Error("fake error for Retry Testing"));
-                Utils.log(e.toString());
+                await().atMost(millis / 2, MILLISECONDS).
+                    untilAsserted(() -> assertEquals(fails, fails - 1));
 
+                RuntimeErrorException e = new RuntimeErrorException(new Error("fake error for Retry Testing"));
+                log(e.toString());
                 // We will countDown the latch in the finally block
-
                 throw e;
             }
 
-            Utils.log("Task " + taskId + " sleeping for " + millis + " milliseconds. " + now
-                    + " workers inside Bulkhead from " + td.getInstances() + " instances " + BAR.substring(0, now));
+            log("Task " + taskId + " sleeping for " + millis + " milliseconds. " + now
+                + " workers inside Bulkhead from " + td.getInstances() + " instances " + BAR.substring(0, now));
 
-            Thread.sleep(millis);
-
-            Utils.log("Task " + taskId + " woke.");
+            await().atMost(millis, MILLISECONDS).untilAsserted(() -> assertTrue(fails < 0));
 
+            log("Task " + taskId + " woke.");
             // We will countDown the latch in the finally block
-
         }
-        catch (InterruptedException e) {
-            Utils.log(e.toString());
+        catch (RuntimeException e) {
+            log(e.toString());
```

Comment:
Is it possible to remove this catch block so that an exception here fails the test?

In general an unexpected exception should fail the test, not be caught and logs. No-one looks at test logs unless the test fails.

Reply:
This try-catch is not possible to remove at this moment due this method has 2 different flows base on how many failed threads the test wants to simulate. I think this can be improved later after review all the other tests that use this method. This is just the first refactoring.